### DISCUSSION TITLE
route:cache on PHP 7.4 Erroneous data format for unserializing

### DISCUSSION THREAD

**Discussion by josezenem** [laravel/framework#31056] [General] [answered]
- Laravel Version: 6.10.0
- PHP Version: 7.4.0 & 7.4.1
- Database Driver & Version: mysql/5.7.28
- laravel/homestead (vmware_desktop, 9.1.0)

### Description:
I was able to replicate this on Homestead and production, I was not able to reproduce this on older PHP version.  The issue is that if you run artisan route:cache it will cache successfully, but at the moment of going to it via a browser it will throw an error: `Erroneous data format for unserializing 'Symfony\Component\Routing\CompiledRoute'`

### Steps To Reproduce:

1. make a new laravel application `laravel new blog`
2. update blog/app⁩/Http⁩/⁨Controllers⁩/Controller.php and replace contents with:
```
<?php

namespace App\Http\Controllers;

use Illuminate\Foundation\Auth\Access\AuthorizesRequests;
use Illuminate\Foundation\Bus\DispatchesJobs;
use Illuminate\Foundation\Validation\ValidatesRequests;
use Illuminate\Routing\Controller as BaseController;

class Controller extends BaseController
{
    use AuthorizesRequests, DispatchesJobs, ValidatesRequests;
    public function home()
    {
    	return 'hello';
    }
}
```
3. edit blog/routes⁩/api.php and comment the file out
4. edit blog/routes⁩/web.php and replace it with:
```
<?php
Route::get('/', [
    'uses' => 'Controller@home',
]);
```
5. run `artisan route:cache`, the success message "Routes cached successfully".  At this point you can go to the app via a web browser and be greeted with a message that says "Erroneous data format for unserializing 'Symfony\Component\Routing\CompiledRoute'"

**Comment by pultho** [comment]
Also happening on 7.3.X on any server we deploy to (Plesk on Debian). We're not able to recreate the problem locally - the exception is handled as warning only.
 
Laravel 6.10.1

<img width="846" alt="image" src="https://user-images.githubusercontent.com/18327485/72238955-a75b9e80-35df-11ea-854d-3caf2f9ccb6d.png">

**Comment by pultho** [comment]
Possible solution was posted on Laracasts: https://laracasts.com/discuss/channels/laravel/laravel-artisan-routecache-causes-erroneous-data-exception?reply=572416

Duplicate: https://github.com/laravel/framework/issues/30979

**Comment by ahmic** [comment]
That error also affects Horizon. If routes are cached supervisor can not start

```
sudo supervisorctl restart all
laravel_horizon:laravel_horizon_00: stopped
laravel_horizon:laravel_horizon_00: ERROR (spawn error)
```

horizon.log
```
In routes.php line 15:

  Erroneous data format for unserializing 'Symfony\Component\Routing\Compiled
  Route'
```

**EDIT:**

A problem occurred because I updated PHP to version 7.4 but my supervisor config for Horizon program contained php7.3

I removed version code and everything works fine now

**Comment by n3omaster** [comment]
Try to delete the routes.php file in cache data. It worked for me.

**Reply by ligne13** [reply]
Worked for me also. thanks.

**Comment by maxiz22** [comment]
> That error also affects Horizon. If routes are cached supervisor can not start
> 
> ```
> sudo supervisorctl restart all
> laravel_horizon:laravel_horizon_00: stopped
> laravel_horizon:laravel_horizon_00: ERROR (spawn error)
> ```
> 
> horizon.log
> 
> ```
> In routes.php line 15:
> 
>   Erroneous data format for unserializing 'Symfony\Component\Routing\Compiled
>   Route'
> ```
> 
> **EDIT:**
> 
> A problem occurred because I updated PHP to version 7.4 but my supervisor config for Horizon program contained php7.3
> 
> I removed version code and everything works fine now

I had same error and fix updating PHP version to 7.4 in supervisor thanks

**Comment by guice** [comment]
We're running within a dockerized environment (k8s), php 7.3.19, Laravel v6.18.7.

We have consumer code built to restart itself ever few hours. Unfortunately, after a period of time, we run into this issue and our consumers stop working with crashlookbackoff.

The given solutions, e.g. delete 'route.php' or delete cache, aren't real solutions for us. k8s is restarting the containers on our behalf. These are patches to the underlining problem -- which is ???. The thing is, most of the time the pods restart just fine. But when they fail, they all fail with the same error. We had 3 pods die last night, all with the same error: 

```
In routes.php line 15:
                                                                               
  Erroneous data format for unserializing 'Symfony\Component\Routing\Compiled  
  Route'                                                                       
                                                                               

About to run 'php artisan optimize'
[2020-07-02 00:19:20] production.ERROR: Erroneous data format for unserializing 'Symfony\Component\Routing\CompiledRoute' {"exception":"[object] (ErrorException(code: 0): Erroneous data format for unserializing 'Symfony\\Component\\Routing\\CompiledRoute' at /app/bootstrap/cache/routes.php:15)
[stacktrace]
```

**Reply by codegeek1001** [reply]
@guice, check your PHP version being used by Laravel vs the docker php. They should be same and if not, this could cause the issue. Another issue that is discussing this possible problem. [Issue # 30979](https://github.com/laravel/framework/issues/30979)

**Comment by GrahamCampbell** [comment, marked as answer]
This happens if you update your dependencies but forget to re-cache the routes. You should also make sure the same version of PHP is running fpm as cli, so your web server PHP version is the same used to generate the routes cache.

**Reply by guice** [reply]
This happens to us when our pods restart - containers, all static code. Definitely no related to any dependency updates.

**Comment by netdjw** [comment]
If I delete the `bootsrap/cache/routes.php` it's working fine. But if I run the `php artisan optimize` again... so it's generate the `routes.php` and again fails.

In my case I use Laravel 6.x on PHP 7.4 on Ubuntu 18

**Reply by GrahamCampbell** [reply]
This only happens if your php-cli is a different version to your php-fpm.

**Reply by netdjw** [reply]
true... the cli was 7.4, fpm was 7.2 ... removed 7.2, and started php7.4-fpm, but still the same error

**Reply by josezenem** [reply]
@netdjw can you double check your fpm via doing a phpinfo() on one of the PHP files, just to ensure the right version did get picked up.

When we ran into the issue a while back, we just had to correct the versions for both CLI and Apache at the time, you may have to also re-run artisan route:cache again.
